.circleci/config.yml

```diff
@@ -1,79 +1,121 @@
+      - save_cache: *save-git-cache
 
-      - save_cache:
-          name: Save git cache
-          key: v1-20180524-git
-          paths:
-            - ".git"
-
-      - run:
-          name: Create Directories for Results and Artifacts
-          command: |
-            mkdir -p $CIRCLE_ARTIFACTS
-            mkdir -p $CIRCLE_TEST_REPORTS
-
-      - restore_cache:
-          name: "Restoring npm cache"
-          keys:
-             - v1-npmcache-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
-             - v1-npmcache-{{ checksum ".nvmrc" }}
-
+      - restore_cache: *restore-npm-cache
       - run: npm ci
+      - save_cache: *save-npm-cache
+      - persist_to_workspace:
+          root: .
+          paths: .
+
+  #
+  # Build translations
+  #
+  translate:
+    docker: *docker-image
+    working_directory: ~/wp-calypso
+    environment: *shared-environment
 
-      - save_cache:
-          name: "Saving npm cache"
-          key: v1-npmcache-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
-          paths:
-            - ~/.npm
+    steps:
+      - attach_workspace:
+          at: .
 
-      - run:
-          name: Build Server
-          environment:
-              NODE_ENV: test
-          command: npm run build-server
+      - run: *setup-results-and-artifacts
 
       - run:
           name: Build calypso-strings.pot
           command: |
-            if [ "$CIRCLE_NODE_INDEX" == "0" ]; then
-              npm run translate; mkdir -p $CIRCLE_ARTIFACTS/translate
-              mv calypso-strings.pot $CIRCLE_ARTIFACTS/translate
-            fi
+            npm run translate; mkdir -p $CIRCLE_ARTIFACTS/translate
+            mv calypso-strings.pot $CIRCLE_ARTIFACTS/translate
 
       - run:
           name: Build New Strings .pot
           command: |
-            if [ "$CIRCLE_NODE_INDEX" == "0" ]; then
-              git clone https://github.com/Automattic/gp-localci-client.git
-              bash gp-localci-client/generate-new-strings-pot.sh $CIRCLE_BRANCH $CIRCLE_SHA1 $CIRCLE_ARTIFACTS/translate
-              rm -rf gp-localci-client
-            fi
+            git clone https://github.com/Automattic/gp-localci-client.git
+            bash gp-localci-client/generate-new-strings-pot.sh $CIRCLE_BRANCH $CIRCLE_SHA1 $CIRCLE_ARTIFACTS/translate
+            rm -rf gp-localci-client
+      - store_artifacts:
+          path: /tmp/artifacts
 
-      - run:
-          name: Run Integration Tests
-          command: |
-            bin/run-integration $(circleci tests glob "bin/**/integration/*.js" "client/**/integration/*.js" "server/**/integration/*.js" | circleci tests split --split-by=timings)
+  #
+  # Test Calypso
+  #
+  test-calypso:
+    parallelism: 6
+    docker: *docker-image
+    working_directory: ~/wp-calypso
+    environment: *shared-environment
+
+    steps:
+      - attach_workspace:
+          at: .
+
+      - run: *setup-results-and-artifacts
 
       - run:
           name: Lint Config Keys
```

Comment:
something like these lint jobs could also be done in parallel. is there any benefit in assigning them to separate jobs vs. using the inherent `parallelism` here?

Reply:
There's some overhead for almost all of our jobs because they all require some setup and teardown (repo checkout, npm install, store test results). There's a base cost for running pretty much any of our jobs.

It's also important to economize containers. We have a limited container pool, and if we use them, other jobs are queued. If setup and teardown takes ~10s and we imagine some short jobs where the principal steps take 2s, both jobs might finish earlier if they're parallelized, but considering overhead it's a net loss:

Parallelized jobs: 10s overhead + 2s steps = 12s * 2 = 24s
Single job: 10s overhead + 2s steps + 2s steps = 14s

If the jobs start at the same moment, the parallel workflow might finish 2s earlier at 12s, but it uses 24s of container time. In the single job, we trade 2s of clock time but free up 10s of container tim.

The lint steps are generally very fast because we're only those that differ from master (#25494). I think there's good reason to keep them together. In #27180, the lint and translation steps are all part of one job.

The other reason to split jobs is that they can be selectively run via the workflow with filters for branches, tags, cron schedules, manual approval. For example, a tag or manual approval workflow might be an interesting idea to explore in #27282

Reply:
In the context of this PR, there's good reason to isolate the client tests from any other steps because the client tests are by far the most time consuming. #27180 does just that.